validator_client/src/initialized_validators.rs

```diff
@@ -401,13 +543,50 @@ impl InitializedValidators {
                 }
             } else {
                 self.validators.remove(&def.voting_public_key);
+                match &def.signing_definition {
+                    SigningDefinition::LocalKeystore {
+                        voting_keystore_path,
+                        ..
+                    } => {
+                        if let Some(key_store) = key_stores.get(voting_keystore_path) {
+                            disabled_uuids.insert(*key_store.uuid());
+                        }
+                    }
+                }
+
                 info!(
                     self.log,
                     "Disabled validator";
                     "voting_pubkey" => format!("{:?}", def.voting_public_key)
                 );
             }
         }
+        for uuid in disabled_uuids {
+            key_cache.remove(&uuid);
+        }
+
+        let validators_dir = self.validators_dir.clone();
+        let log = self.log.clone();
+        if key_cache.is_modified() {
+            executor.spawn_blocking(
```

Comment:
The `v0.3.0-staging` branch has this `update_validators` function, so we can await this `spawn_blocking` and avoid some race conditions.